node/subsystem/src/lib.rs

```diff
@@ -14,366 +14,77 @@
-		where T: IntoIterator<Item = AllMessages> + Send, T::IntoIter: Send
-	{
-		self.sender().send_messages(msgs).await
-	}
+impl<T> SubsystemSender for T where T: polkadot_overseer::gen::SubsystemSender<messages::AllMessages> {
+}
 
+/// Spawned subsystem.
+pub type SpawnedSubsystem = polkadot_overseer::gen::SpawnedSubsystem<SubsystemError>;
 
-	/// Send a message onto the unbounded queue of some other `Subsystem`, routed based on message
-	/// type.
-	///
-	/// This function should be used only when there is some other bounding factor on the messages
-	/// sent with it. Otherwise, it risks a memory leak.
-	///
-	/// Generally, for this method to be used, these conditions should be met:
-	/// * There is a communication cycle between subsystems
-	/// * One of the parts of the cycle has a clear bound on the number of messages produced.
-	fn send_unbounded_message(&mut self, msg: AllMessages) {
-		self.sender().send_unbounded_message(msg)
-	}
-}
 
-/// A trait that describes the [`Subsystem`]s that can run on the [`Overseer`].
-///
-/// It is generic over the message type circulating in the system.
-/// The idea that we want some type contaning persistent state that
-/// can spawn actually running subsystems when asked to.
-///
-/// [`Overseer`]: struct.Overseer.html
-/// [`Subsystem`]: trait.Subsystem.html
-pub trait Subsystem<C: SubsystemContext> {
-	/// Start this `Subsystem` and return `SpawnedSubsystem`.
-	fn start(self, ctx: C) -> SpawnedSubsystem;
+/// Convenience trait specialization.
+pub trait SubsystemContext: polkadot_overseer::gen::SubsystemContext<
+	Signal=OverseerSignal,
+	AllMessages=messages::AllMessages,
+	Error=SubsystemError,
+>
+{
+	/// The message type the subsystem consumes.
+	type Message: std::fmt::Debug + Send + 'static;
+	/// Sender type to communicate with other subsystems.
+	type Sender: SubsystemSender + Send + Clone + 'static;
 }
 
-/// A dummy subsystem that implements [`Subsystem`] for all
-/// types of messages. Used for tests or as a placeholder.
-pub struct DummySubsystem;
-
-impl<C: SubsystemContext> Subsystem<C> for DummySubsystem
+impl<T> SubsystemContext for T
 where
-	C::Message: std::fmt::Debug
+	T: polkadot_overseer::gen::SubsystemContext<
+		Signal=OverseerSignal,
+		AllMessages=messages::AllMessages,
+		Error=SubsystemError,
+	>,
 {
-	fn start(self, mut ctx: C) -> SpawnedSubsystem {
-		let future = Box::pin(async move {
-			loop {
-				match ctx.recv().await {
-					Err(_) => return Ok(()),
-					Ok(FromOverseer::Signal(OverseerSignal::Conclude)) => return Ok(()),
-					Ok(overseer_msg) => {
-						tracing::debug!(
-							target: "dummy-subsystem",
-							"Discarding a message sent from overseer {:?}",
-							overseer_msg
-						);
-						continue;
-					}
-				}
-			}
-		});
-
-		SpawnedSubsystem {
-			name: "dummy-subsystem",
-			future,
-		}
-	}
+	type Message = <Self as polkadot_overseer::gen::SubsystemContext>::Message;
+	type Sender = <Self as polkadot_overseer::gen::SubsystemContext>::Sender;
```

Comment:
have you tried instead do
```rust
trait SubsystemContext {
  type Message;
  ...
}
impl<T: SubsystemContext> overseer::SubsystemContext for T {
  types AllMessages = AllMessages;
  ...
}
```
Would it solve the problem of having to specify two traits?

Reply:
The issue will stay, since the the information of the associated types for which we impl the `overseer::SubsystemContext` will not propate to the `fn foo(ctx: impl SubsystemContext)` and hence we can not use concrete types there, which is what we want..

Reply:
That was indeed a variation I did not try before, but the `overseer::SubsystemContext` is not defined in the same crate, and hence `impl<T> X for T {}` will not work either:

```
error[E0210]: type parameter `T` must be used as the type parameter for some local type (e.g., `MyStruct<T>`)
   --> node/subsystem/src/lib.rs:110:6
    |
110 | impl<T> polkadot_overseer::gen::SubsystemContext for T
    |      ^ type parameter `T` must be used as the type parameter for some local type
    |
    = note: implementing a foreign trait is only possible if at least one of the types for which it is implemented is local
    = note: only traits defined in the current crate can be implemented for a type parameter

error: aborting due to previous error
```

Reply:
As a conclusion there are two options:

* accept as is
* generate the traits with the proper types and avoid the clutter

Reply:
I don't have a strong opinion. Less clutter is good, but if that's cumbersome to achieve, I think it is fine.